src/client/java/daos-java-api/src/main/native/com_intel_daos_client_DaosFsClient.c

```diff
@@ -0,0 +1,911 @@
+		sprintf(msg, tmp, file_path, flags);
+		throw_exception(env, msg, rc);
+		file = NULL;
+	}
+	(*env)->ReleaseStringUTFChars(env, path, file_path);
+	return *(jlong*)&file;
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsGetSize
+  (JNIEnv *env, jobject client, jlong dfsPtr, jlong objId){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	dfs_obj_t *file = *(dfs_obj_t**)&objId;
+	daos_size_t size;
+	int rc = dfs_get_size(dfs, file, &size);
+	if(rc){
+		char *tmp = "Failed to get file size";
+		throw_exception_const_msg(env, tmp, rc);
+		return -1;
+	}
+	return size;
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsDup
+  (JNIEnv *env, jobject client, jlong dfsPtr, jlong objId, jint flags){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	dfs_obj_t *file = *(dfs_obj_t**)&objId;
+	dfs_obj_t *new_file;
+	int rc = dfs_dup(dfs, file, flags, &new_file);
+	if (rc) {
+		char *tmp = "Failed to duplicate file";
+		throw_exception_const_msg(env, tmp, rc);
+		return -1;
+	}
+	return *(jlong*)&new_file;
+}
+
+JNIEXPORT void JNICALL Java_com_intel_daos_client_DaosFsClient_dfsRelease
+  (JNIEnv *env, jclass clientClass, jlong objId){
+	dfs_obj_t *file = *(dfs_obj_t**)&objId;
+	int rc = dfs_release(file);
+	if (rc) {
+		char *tmp = "Failed to release file";
+		throw_exception_const_msg(env, tmp, rc);
+	}
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsRead
+  (JNIEnv *env, jobject client, jlong dfsPtr, jlong objId, jlong bufferAddress, jlong fileOffset, jlong len, jint eventNo){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	dfs_obj_t *file = *(dfs_obj_t**)&objId;
+	char *buf = (char*)bufferAddress;
+	d_iov_t sg_iov = {0};
+	d_sg_list_t sgl = {
+		.sg_nr = 1,
+		.sg_nr_out = 0,
+		.sg_iovs = &sg_iov
+	};
+	d_iov_set(&sg_iov, buf, len);
+	daos_size_t size = 0;
+	int rc = dfs_read(dfs, file, &sgl, fileOffset, &size, NULL);
+	if (rc) {
+		char *tmp = "Failed to read %ld bytes from file starting at %ld";
+		char *msg = (char *)malloc(strlen(tmp) + 8 + 8);
+		sprintf(msg, tmp, len, fileOffset);
+		throw_exception(env, msg, rc);
+		return 0;
+	}
+	return size;
+}
+
+JNIEXPORT jlong JNICALL Java_com_intel_daos_client_DaosFsClient_dfsWrite
+  (JNIEnv *env, jobject client, jlong dfsPtr, jlong objId, jlong bufferAddress, jlong fileOffset, jlong len, jint eventNo){
+	dfs_t *dfs = *(dfs_t**)&dfsPtr;
+	dfs_obj_t *file = *(dfs_obj_t**)&objId;
+	char *buf = (char*)bufferAddress;
+	d_iov_t sg_iov = {0};
+	d_sg_list_t sgl = {
+		.sg_nr = 1,
+		.sg_nr_out = 0,
+		.sg_iovs = &sg_iov
+	};
+	d_iov_set(&sg_iov, buf, len);
+	int rc = dfs_write(dfs, file, &sgl, fileOffset, NULL);
+	if (rc) {
+		char *tmp = "Failed to write %ld bytes to file starting at %ld";
+		char *msg = (char *)malloc(strlen(tmp) + 8 + 8);
+		sprintf(msg, tmp, len, fileOffset);
+		throw_exception(env, msg, rc);
+		return 0;
+	}
+	return len;
+}
+
+//TODO: support max entries
+JNIEXPORT jstring JNICALL Java_com_intel_daos_client_DaosFsClient_dfsReadDir
```

Comment:
shouldn't you try and return something like the dirent struct as an object?

Reply:
For now, we don't need it. We can provide dirent java corresponding in later release.